core/fetch/bootstrap.go

```diff
@@ -0,0 +1,206 @@
+package fetch
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"chain/core/rpc"
+	"chain/core/txdb"
+	"chain/errors"
+	"chain/protocol"
+	"chain/protocol/bc"
+)
+
+// SnapshotProgress describes a snapshot being downloaded from a peer Core.
+type SnapshotProgress struct {
+	mu               sync.Mutex
+	attempt          int
+	height           uint64
+	size             uint64
+	downloadProgress *progressReader
+
+	stopped chan struct{}
+}
+
+// Attempt returns the how many times Core has attempted to
+// download a bootstrap snapshot. If a download request times out
+// or encounters any kind of validation error, it'll re-attempt the
+// snapshot download a few times.
+func (s *SnapshotProgress) Attempt() int {
```

Comment:
These accessors should acquire the mutex before reading `SnapshotProgress` fields.